modules/standard/Random.chpl

```diff
@@ -915,25 +1124,87 @@ module Random {
 
       /* Generate a random number in [0,bound).
 
+         This function corresponds to pcg32_boundedrand_r and can
+         call the random-number generator more than once.
+
+         :arg inc: The sequence constant (same as passed to `srandom`)
+         :arg bound: The returned value will be < `bound`.
+         :returns: a random number in [0,bound).
+       */
+      inline
+      proc bounded_random(inc:uint(64), bound:uint(32))
+      {
+            // This comment is from pcg32_boundedrand_r:
+            // To avoid bias, we need to make the range of the RNG a multiple of
+            // bound, which we do by dropping output less than a threshold.
+            // A naive scheme to calculate the threshold would be to do
+            //
+            //     uint32_t threshold = 0x100000000ull % bound;
+            //
+            // but 64-bit div/mod is slower than 32-bit div/mod (especially on
+            // 32-bit platforms).  In essence, we do
+            //
+            //     uint32_t threshold = (0x100000000ull-bound) % bound;
+            //
+            // because this version will calculate the same modulus, but the LHS
+            // value is less than 2^32.
+
+        const negbound:uint(32) = ( -(bound:int(32)) ):uint(32);
+        const threshold:uint(32) = negbound % bound;
+        var tmprng:pcg_setseq_64_xsh_rr_32_rng;
+        var tmpinc:uint(64);
+        var r:uint(32);
+
+        // Keepy trying until we get a random number that is within the bounds.
+        while true {
+          r = random(inc);
+          if r >= threshold then
+            return r % bound;
+        }
+
+        // never reached.
+        return 0;
+      }
+
+
+      /* Generate a random number in [0,bound).
+
          This function corresponds to pcg32_boundedrand_r, but has one
          difference. Because parallel random number generation relies
```

Comment:
bounded_random_vary_inc only runs this.random once. The other random calls are to temporary RNGs and so don't advance the original RNG further.